src/platform/ESP32/MdnsImpl.cpp

```diff
@@ -0,0 +1,241 @@
+    SuccessOrExit(error = DeviceLayer::ConfigurationMgr().GetSetupDiscriminator(discriminator));
+    snprintf(hostname, sizeof(hostname), "CHIP-%04d", discriminator);
+    espError = mdns_hostname_set(hostname);
+    VerifyOrExit(espError == ESP_OK, error = CHIP_ERROR_INTERNAL);
+
+exit:
+    if (espError != ESP_OK)
+    {
+        ChipLogError(DeviceLayer, "esp mdns internal error: %s", esp_err_to_name(espError));
+    }
+    initCallback(context, error);
+
+    return error;
+}
+
+static const char * GetProtocolString(MdnsServiceProtocol protocol)
+{
+    return protocol == MdnsServiceProtocol::kMdnsProtocolTcp ? "_tcp" : "_udp";
+}
+
+CHIP_ERROR ChipMdnsPublishService(const MdnsService * service)
+{
+    CHIP_ERROR error        = CHIP_NO_ERROR;
+    mdns_txt_item_t * items = nullptr;
+    esp_err_t espError;
+
+    VerifyOrExit(service->mTextEntrySize <= UINT8_MAX, error = CHIP_ERROR_INVALID_ARGUMENT);
+    if (service->mTextEntryies)
+    {
+        items = static_cast<mdns_txt_item_t *>(chip::Platform::MemoryCalloc(service->mTextEntrySize, sizeof(mdns_txt_item_t)));
+        VerifyOrExit(items != nullptr, error = CHIP_ERROR_NO_MEMORY);
+        for (size_t i = 0; i < service->mTextEntrySize; i++)
+        {
+            items[i].key = service->mTextEntryies[i].mKey;
+            // Unfortunately ESP mdns stack dosen't support arbitrary binary data
+            items[i].value = reinterpret_cast<const char *>(service->mTextEntryies[i].mData);
+        }
+    }
+
+    espError = mdns_service_add(service->mName, service->mType, GetProtocolString(service->mProtocol), service->mPort, items,
+                                service->mTextEntrySize);
+    // The mdns_service_add will return error if we try to add an existing service
+    if (espError != ESP_OK)
+    {
+        espError = mdns_service_txt_set(service->mType, GetProtocolString(service->mProtocol), items,
+                                        static_cast<uint8_t>(service->mTextEntrySize));
+    }
+    VerifyOrExit(espError == ESP_OK, error = CHIP_ERROR_INTERNAL);
+
+exit:
+    if (items != nullptr)
+    {
+        chip::Platform::MemoryFree(items);
+    }
+
+    return error;
+}
+
+CHIP_ERROR ChipMdnsStopPublish()
+{
+    return mdns_service_remove_all() == ESP_OK ? CHIP_NO_ERROR : CHIP_ERROR_INTERNAL;
+}
+
+CHIP_ERROR ChipMdnsBrowse(const char * type, MdnsServiceProtocol protocol, chip::Inet::InterfaceId interface,
+                          MdnsBrowseCallback callback, void * context)
+{
+    mdns_result_t * results = nullptr;
+    mdns_result_t * current;
+    CHIP_ERROR error = CHIP_NO_ERROR;
+    esp_err_t espError;
+    MdnsService * services = nullptr;
+    TextEntry * textEntries;
+    size_t serviceSize = 0;
+
+    ChipLogProgress(DeviceLayer, "call mdns_query_ptr, type=%s protocol=%s", type, GetProtocolString(protocol));
+    espError = mdns_query_ptr(type, GetProtocolString(protocol), kTimeoutMilli, kMaxResults, &results);
```

Comment:
This appears blocking and will block for a huge time. If this is run from main protocol loop, this will likely break the rpotocol, or cause packetbuffer backup from Wifi down to CHIP stack that will may cause memory exhaustion.

Also, kMaxResult of 20 is not very large. I strongly suggest avoiding the browsing altogether on embedded, apart from debugging.

Reply:
This API is just implemented as "better than nothing". We only publish in the main loop. Do we prefer to return `CHIP_ERROR_NOT_IMPLEMENTED`?

Reply:
I definitely think that for embedded, CHIP_ERROR_NOT_IMPLEMENTED is OK for browsing. The embedded devices are limited enough that browsing is likely a problem and this was called out in the SCTT discovery PR.